code42/py42
# Feature/integ 1737 xfc support

### Description of Change ###

Allows customers that do not use Code42 preservations feature to use the `sdk.securitydata.stream_file_by_md5()` and `sdk.securitydata.stream_file_by_sha256()` functions to stream files that were captured and via an agent's exfiltration detection.

This means that a given search could pull its file from either preservation data or exfiltration data. The logic to figure out which it should is as follows:
- if there is an exact match via exfiltration data, use this.
- if there is a path match via exfiltration data, use the latest version that matches the path
- otherwise, follow the same preservation logic as previously -- check preservation data for an exact match, if there isn't one, use the latest version.

Also refactors the usage of `py42_response` in the test suite. This makes setting up mock responses much easier and more consistently mimics their behavior.

### Issues Resolved ###

- closes INTEG-1737

### Testing Procedure ###
Call both `sdk.securitydata.stream_file_by_md5()` and `sdk.securitydata.stream_file_by_sha256()` in each of the following circumstances and ensure that a file is successfully downloaded:
- hash is available via preservation data 
- hash has an exact match
- hash has a matching path
- hash falls back to searching other devices that have this hash, and is available via preservation data of one of _those_ devices.

There is a tenant in our core-int environment that has suitable events for testing, slack me for details.

### PR Checklist ###
Did you remember to do the below?
